### DISCUSSION TITLE
custom_scripts show other value in a selectbox for ObjectVar

### DISCUSSION THREAD

**Discussion by Frank17489** [netbox-community/netbox#14447] [Help Wanted!]
Hi,

I'am new in scripting for netbox. 

My task is to create a new device. The user can select a network (prefix). If the device gets a LAN interface (preset in the DeviceType), then a free IP address from the selected prefix should be created and assigned to the interface.

My questions:
1. How I can program, that the description be displayed instead of the prefix when user selecting a prefix?
2. how can I get a free IP address of the specific prefix in the script?

thx for any advice

**Comment by Frank17489** [comment]
for 2. give chatGPT the correct answer:

```
# NetBox API configuration
base_url = "https://netbox.example.com/api/"
token = "your_api_token_here"
headers = {"Authorization": f"Token {token}"}
prefix_id = 1  # Replace with the actual ID of the prefix you are interested in

# API endpoint for IP addresses within a prefix
endpoint = f"ipam/prefixes/{prefix_id}/available-ips/"

# Make the GET request
response = requests.get(f"{base_url}{endpoint}", headers=headers)

# Check the status code and process the response
if response.status_code == 200:
    available_ips = response.json()
    
    if available_ips:
        # Get the first available IP address
        new_ip_address = available_ips[0]["address"]
        print(f"New IP Address: {new_ip_address}")
    else:
        print("No available IP addresses in the specified prefix.")
else:
    print(f"Error: {response.status_code} - {response.text}")
```

**Reply by candlerb** [reply]
Don't trust what ChatGPT says. That answer uses the REST API, but in Custom Scripts you should use the ORM directly. That way you gain transaction safety and performance, as well as not needing credentials to call back to the REST API.

I'll take the simpler question first:

> how can I get a free IP address of the specific prefix in the script?

In other words, "how can I use the ORM to pick the first available IP address in a prefix?"

You can get the answer either by looking at the [model source](https://github.com/netbox-community/netbox/blob/v3.6.6/netbox/ipam/models/ip.py#L444-L451), or by running nbshell and typing `help(Prefix)`. The answer is:

```
ip = pfx.get_first_available_ip()    # returns "x.x.x.x/y" as a string, or None if there are none available
```

e.g.

```
# /opt/netbox/venv/bin/python /opt/netbox/netbox/manage.py nbshell
### NetBox interactive shell (netbox3)
### Python 3.8.10 | Django 4.2.7 | NetBox 3.6.6
### lsmodels() will show available models. Use help(<model>) for more info.
>>> pfx = Prefix.objects.get(prefix="10.12.0.0/24")
>>> pfx.get_first_available_ip()
'10.12.0.2/24'
>>> help(Prefix)
...
 |  get_available_ips(self)
 |      Return all available IPs within this prefix as an IPSet.
...
 |  get_first_available_ip(self)
 |      Return the first available IP within the prefix (or None).
```

Beware: if there's any chance that two copies of the same script could be running at the same time, you should use advisory locks to ensure they can't both pick the same IP address.

> How I can program, that the description be displayed instead of the prefix when user selecting a prefix?

That's harder. According to discussion #10047 there used to be `display_field` but that's no longer available (widget is apparently hard-coded to use the "display" attribute in the REST API response), and #6018 says this is not possible.

You could always take the entire list of prefixes of interest and materialize them into selections for a [ChoiceVar](https://docs.netbox.dev/en/stable/customization/custom-scripts/#choicevar). But if there's a large number of prefixes then this would be slow and expensive.

Otherwise, according to [Django documentation](https://docs.djangoproject.com/en/4.2/ref/forms/fields/#modelchoicefield):

> The __str__() method of the model will be called to generate string representations of the objects for use in the field’s choices. To provide customized representations, subclass ModelChoiceField and override label_from_instance. This method will receive a model object and should return a string suitable for representing it. For example: ...

I don't see any mention of "label_from_instance" in Netbox discussions or issues.  It might be possible to subclass ModelChoiceField yourself; for plugin models I see this done in #9929. Let us know if you get a solution.

**Reply by Frank17489** [reply]
Hi candlerb,

thx for your advices. ChatGPT was my last try. Now you give me new ideas.
To modifying  the model is in moment too difficult for me.
